qutebrowser/utils/psl.py

```diff
@@ -0,0 +1,145 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2014-2021 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+# Copyright 2021 Javier Ayres <javier@lufte.net>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <https://www.gnu.org/licenses/>.
+
+import enum
+import os.path
+
+from PyQt5.QtCore import QUrl
+
+from qutebrowser.utils import standarddir, urlutils
+
+
+class PSLNotFound(Exception):
+    """Raised when no public_suffix_list.dat file is found.
+
+    Attributes:
+        paths: iterable of all tried paths.
+    """
+
+    def __init__(self, paths):
+        super().__init__(
+            "Couldn't find public_suffix_list.dat in any of the expected paths: "
+            "{}".format(', '.join(paths))
+        )
+
+
+class _Rule(enum.Enum):
+
+    NORMAL = enum.auto()
+    WILDCARD = enum.auto()
+    EXCEPTION = enum.auto()
+
+
+def _widened_hostnames(hostname):
+    """A generator for widening string hostnames.
+
+    Ex: a.c.foo -> [a.c.foo, c.foo, foo]"""
+    while hostname:
+        yield hostname
+        hostname = hostname.partition(".")[-1]
+
+
+class _SuffixList:
+
+    _rules = None
+
+    def _load_rules(self) -> None:
+        PSL_PATHS = (
+            # User-provided list in data directory (need to wait until standarddir is
+            # initialized)
+            os.path.join(standarddir.data(), 'public_suffix_list.dat'),
+            # Bundled list
+            os.path.join(__file__, '../public_suffix_list.dat'),
```

Comment:
Where would a bundled list live?

Reply:
You'd need to read it with `utils.read_file('3rdparty/public_suffix_list.dat')` or so. That will then return the file contents via Python's [resource system](https://docs.python.org/3/library/importlib.html#module-importlib.resources), which might or might not be at a filesystem path (it could also be inside a .zip-like Python egg, depending on how qutebrowser is installed).

This would probably make the logic a bit more complicated (as you can't just go through a list like here), but I don't see a way around this.